opentelemetry_collector/receiver/vmagereceiver/vm_age_collector.go

```diff
@@ -99,40 +119,55 @@ func (collector *VMAgeCollector) StopCollection() {
 }
 
 func (collector *VMAgeCollector) makeErrorMetrics() *metricspb.Metric {
-	timeseries := metricgenerator.MakeInt64TimeSeries(1, collector.startTime, time.Now(), collector.labelValues)
+	timeseries := metricgenerator.MakeInt64TimeSeries(1, collector.collectorStartTime, time.Now(), collector.labelValues)
 	return &metricspb.Metric{
-		MetricDescriptor: vmImageErrorMetric,
+		MetricDescriptor: vmImageAgesErrorMetric,
 		Timeseries:       []*metricspb.TimeSeries{timeseries},
 	}
+}
+
+func makeMetrics(metricDescriptor *metricspb.MetricDescriptor, timeseries *metricspb.TimeSeries) []*metricspb.Metric {
+	return []*metricspb.Metric{
+		{
+			MetricDescriptor: metricDescriptor,
+			Timeseries:       []*metricspb.TimeSeries{timeseries},
+		},
+	}
+}
 
+func (collector *VMAgeCollector) export(metrics []*metricspb.Metric, errorKey string) {
+	ctx := context.Background()
+	md := consumerdata.MetricsData{Metrics: metrics}
+	err := collector.consumer.ConsumeMetrics(ctx, internaldata.OCSliceToMetrics([]consumerdata.MetricsData{md}))
+	if err != nil {
+		collector.logger.Error(errorKey, zap.Error(err))
+	}
 }
 
-func (collector *VMAgeCollector) scrapeAndExport() {
-	metrics := make([]*metricspb.Metric, 0, 1)
+func (collector *VMAgeCollector) scrapeAndExportVMImageAge() {
+	var metrics []*metricspb.Metric
 
 	if collector.buildDateError {
-		metrics = append(metrics, collector.makeErrorMetrics())
+		metrics = []*metricspb.Metric{collector.makeErrorMetrics()}
 	} else {
 		imageAge, err := calculateImageAge(collector.parsedBuildDate, time.Now())
 		if err != nil {
-			metrics = append(metrics, collector.makeErrorMetrics())
+			metrics = []*metricspb.Metric{collector.makeErrorMetrics()}
 		} else {
-			timeseries := metricgenerator.MakeDoubleTimeSeries(
-				imageAge, collector.startTime, time.Now(), collector.labelValues)
-			metrics = append(
-				metrics,
-				&metricspb.Metric{
-					MetricDescriptor: vmImageAgeMetric,
-					Timeseries:       []*metricspb.TimeSeries{timeseries},
-				},
-			)
+			timeseries := metricgenerator.MakeDoubleTimeSeries(imageAge, collector.collectorStartTime, time.Now(), collector.labelValues)
+			metrics = makeMetrics(vmImageAgeMetric, timeseries)
 		}
 	}
 
-	ctx := context.Background()
-	md := consumerdata.MetricsData{Metrics: metrics}
-	err := collector.consumer.ConsumeMetrics(ctx, internaldata.OCSliceToMetrics([]consumerdata.MetricsData{md}))
-	if err != nil {
-		collector.logger.Error("Error sending metrics", zap.Error(err))
+	collector.export(metrics, "Error sending VM image age metrics")
+}
+
+func (collector *VMAgeCollector) scrapeAndExportVMReadyTime() {
+	if collector.vmReadyTimeError {
```

Comment:
Since error metrics aren't being exported for this metric, this error checking can be done earlier and once. ie don't start the scrapeAndExportVMReadyTime loop at all if there is a vmReadyTimeError.

Reply:
Could you clarify a bit more what you mean?
1. Do you mean to move the "if" into the goroutine? To me, that seems like the same thing.
2. Do you mean to create two separate goroutines and calling the goroutine that does not include "collector.scrapeAndExportVMReadyTime" if there is an error? That seems a bit messy to me for a small optimization.

Reply:
I was thinking of number 2, but you're right it's not really necessary.